File: app/src/main/java/com/gasparaiciukas/owntrainer/activity/AddMealToDiaryActivity.java
package com.gasparaiciukas.owntrainer.activity;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;

import com.gasparaiciukas.owntrainer.R;
import com.gasparaiciukas.owntrainer.adapter.MealAdapter;
import com.gasparaiciukas.owntrainer.database.Meal;
import com.google.android.material.textfield.TextInputEditText;

import java.util.List;

import io.realm.Realm;

public class AddMealToDiaryActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private MealAdapter adapter;
    private RecyclerView.LayoutManager layoutManager;
    private String primaryKey;
    private Realm realm;
    private TextInputEditText portionSizeInput;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_meal_to_diary);

        primaryKey = getIntent().getStringExtra("primaryKey");

        // Get meals from database
        realm = Realm.getDefaultInstance();
        List<Meal> meals = realm.where(Meal.class).findAll();

        // Set up recycler view
        recyclerView = findViewById(R.id.add_meal_to_diary_recycler_view);
        adapter = new MealAdapter(3, meals, primaryKey);
        layoutManager = new LinearLayoutManager(this);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(layoutManager);
    }

    @Override
    protected void onResume() {
        super.onResume();
        adapter.reload();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        realm.close();
    }
}

File: app/src/main/java/com/gasparaiciukas/owntrainer/activity/AddMealToDiaryActivity.kt
package com.gasparaiciukas.owntrainer.activity

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.gasparaiciukas.owntrainer.adapter.MealAdapter
import com.gasparaiciukas.owntrainer.database.DiaryEntry
import com.gasparaiciukas.owntrainer.database.Meal
import com.gasparaiciukas.owntrainer.databinding.ActivityAddMealToDiaryBinding
import io.realm.Realm

class AddMealToDiaryActivity : AppCompatActivity() {
    private lateinit var binding: ActivityAddMealToDiaryBinding
    private lateinit var adapter: MealAdapter
    private lateinit var layoutManager: RecyclerView.LayoutManager
    private lateinit var primaryKey: String
    private lateinit var realm: Realm
    private val listener: (meal: Meal, position: Int) -> Unit = { meal: Meal, position: Int ->
        val realm = Realm.getDefaultInstance()
        val diaryEntry = realm.where(DiaryEntry::class.java)
            .equalTo("yearAndDayOfYear", primaryKey)
            .findFirst()
        if (diaryEntry != null) {
            val mealList = diaryEntry.meals
            realm.executeTransaction {
                mealList.add(meal)
                diaryEntry.meals = mealList
            }
        }
        realm.close()
        finish()
    }
    //private val portionSizeInput: TextInputEditText? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAddMealToDiaryBinding.inflate(layoutInflater)
        setContentView(binding.root)
        primaryKey = intent.getStringExtra("primaryKey").toString()

        // Get meals from database
        realm = Realm.getDefaultInstance()
        val meals: List<Meal> = realm.where(Meal::class.java).findAll()

        // Set up recycler view
        adapter = MealAdapter(meals, listener)
        layoutManager = LinearLayoutManager(this)
        binding.recyclerView.adapter = adapter
        binding.recyclerView.layoutManager = layoutManager
    }

    override fun onResume() {
        super.onResume()
        adapter.reload()
    }

    override fun onDestroy() {
        super.onDestroy()
        realm.close()
    }
}